cellbender/remove_background/vae/encoder.py

```diff
@@ -107,184 +128,207 @@ def forward(self, x: torch.Tensor, **kwargs) -> Dict[str, torch.Tensor]:
         # Initialize p to be a sigmoid function of UMI counts.
         with torch.no_grad():
-            self.output.weight[0][0] = self.INITIAL_OUTPUT_SCALE_FOR_LOG_COUNTS
-            self.output.bias.data.copy_(torch.tensor([self.INITIAL_OUTPUT_BIAS_FOR_LOG_COUNTS
-                                                      * log_count_crossover]))
+            self.output.weight[0][0] = self.INITIAL_WEIGHT_FOR_LOG_COUNTS
+            # Prevent a negative weight from starting something inverted.
+            self.output.weight[1][0] = torch.abs(self.output.weight[1][0])
+            self.output.weight[2][0] = torch.abs(self.output.weight[2][0])
 
         # Set up the non-linear activations.
+        self.nonlin = nn.Softplus()
         self.softplus = nn.Softplus()
 
+        # Set up the initial biases.
+        self.offset = None
+
+        # Set up the initial scaling for values of x.
+        self.x_scaling = None
+
+        # Set up initial values for overlap normalization.
+        self.overlap_mean = None
+        self.overlap_std = None
+
+    def _poisson_log_prob(self, lam, value):
+        return (lam.log() * value) - lam - (value + 1).lgamma()
+
     def forward(self,
                 x: torch.Tensor,
-                chi_ambient: torch.Tensor) -> torch.Tensor:
+                chi_ambient: Union[None, torch.Tensor],
+                z: torch.Tensor,
+                **kwargs) -> Dict[str, torch.Tensor]:
         # Define the forward computation to go from gene expression to cell
         # probabilities.  The log of the total UMI counts is concatenated with
         # the input gene expression and the estimate of the difference between
         # the ambient RNA profile and this barcode's gene expression to form
         # an augmented input.
 
-        # Transform input and calculate log total UMI counts per barcode.
-        x = x.reshape(-1, self.input_dim)
-        log_sum = x.sum(dim=-1, keepdim=True).log1p()
+        x = x.reshape(-1, self.n_genes)
+
+        # Calculate log total UMI counts per barcode.
+        counts = x.sum(dim=-1, keepdim=True)
+        log_sum = counts.log1p()
+
+        # Calculate the log of the number of nonzero genes.
+        log_nnz = (x > 0).sum(dim=-1, keepdim=True).float().log1p()
+
+        # Calculate a similarity between expression and ambient.
+        if chi_ambient is not None:
+            overlap = self._poisson_log_prob(lam=counts * chi_ambient.detach().unsqueeze(0),
+                                             value=x).sum(dim=-1, keepdim=True)
+            if self.overlap_mean is None:
+                self.overlap_mean = (overlap.max() + overlap.min()) / 2
+                self.overlap_std = overlap.max() - overlap.min()
+            overlap = (overlap - self.overlap_mean) / self.overlap_std * 5
+        else:
+            overlap = torch.zeros_like(counts)
+
+        # Apply transformation to data.
         x = transform_input(x, self.transform)
 
+        # TODO: is this helpful?
```

Comment:
A good rule of thumb is to either accompany a `TODO` with github issue number, or not have it at all.

Reply:
Noted, and I see I missed this one... I meant to remove it